### PR TITLE
Implement skill cards

### PR DESCRIPTION
**Pull request by burieberry**
#CS-6757
- Added `skill-card.gts` in base, along with its catalog entry. added `skills` field in room card.
- Added SkillMenu component which now appears in the AI Assistant Panel
- Currently skills are not persisted. Created CS-6985 for this and added 2 skipped tests in matrix.
- First pass on having skills be functional: Skill cards that are attached (and not disabled) are sent to the bot with the sent message. Then the bot responds accordingly. 

Example: 
Before and after adding 'Talk Like a Pirate' skill:
pirate-skill

### CODE REVIEW

**Comment by habdelra** [comment]
Before I do the review could you summarize the goal of this PR and perhaps what linear ticket is pertains to?

**Comment by burieberry** [comment]
> Before I do the review could you summarize the goal of this PR and perhaps what linear ticket is pertains to?

updated the PR description

**Comment by backspace** [comment]
Running this locally, this is maybe beyond the scope of the PR, but it textually proposed changing two fields but only used one in the patch:

Boxel 2024-07-12 14-06-51

Boxel 2024-07-12 14-07-54

Will now look at the code!

File: packages/base/room.gts [hunk 1]
```diff
@@ -461,6 +462,12 @@ export class RoomField extends FieldDef {
     },
   });
 
+  @field skills = containsMany(SkillField);
+
+  addSkill(card: SkillCard) {
+    this.skills.push(new SkillField({ card, isActive: true }));
+  }
+
```
Comment (backspace) [inline packages/base/room.gts:470]: How much adaptation might be needed with the imminence of #1407?

File: packages/base/skill-card.gts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    <template>
+      <@fields.title />
+    </template>
+  };
+}
+
+export class SkillField extends FieldDef {
+  @field isActive = contains(BooleanField);
+  @field card = linksTo(SkillCard);
+}
```
Comment (backspace) [inline packages/base/skill-card.gts:25]: I’m commenting in a semi-random place, looking at this made me wonder about caching; I tried changing the pirate skill card, specifically I said that the terminating phrase should be different, but it didn’t change, even when I turned the skill off and on. But it did change when I refreshed. This is probably not a surprise, but is there a plan to ensure skills are somehow reevaluated when applied ones change?

File: packages/host/app/components/ai-assistant/skill-menu/index.gts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+      }
+      :global(.skill-menu.pill-menu--minimized:hover .expandable-header-button),
+      :global(.skill-menu.pill-menu--minimized:hover .skills-length),
+      :global(
+          .skill-menu.pill-menu--minimized:focus-within
+            .expandable-header-button
+        ),
+      :global(.skill-menu.pill-menu--minimized:focus-within .skills-length) {
+        visibility: visible;
+      }
```
Comment (backspace) [inline packages/host/app/components/ai-assistant/skill-menu/index.gts]:
```suggestion
      .skill-menu.pill-menu--minimized {
        --boxel-pill-menu-width: 3.75rem;
        white-space: nowrap;
        transition: width 0.2s ease-in;
      }
      .skill-menu.pill-menu--minimized:focus {
        outline: 0;
      }
      .skill-menu.pill-menu--minimized:hover,
      .skill-menu.pill-menu--minimized:focus-within {
        --boxel-pill-menu-width: 100%;
      }
      .skill-menu.pill-menu--minimized :deep(.expandable-header-button),
      .skill-menu.pill-menu--minimized :deep(.skills-length) {
        visibility: collapse;
        transition: visibility 0.2s ease-in;
      }
      .skill-menu.pill-menu--minimized:hover :deep(.expandable-header-button),
      .skill-menu.pill-menu--minimized:hover :deep(.skills-length),
      .skill-menu.pill-menu--minimized:focus-within
        :deep(.expandable-header-button),
      .skill-menu.pill-menu--minimized:focus-within :deep(.skills-length) {
        visibility: visible;
      }
```

I was able to bypass needing `:global` without any visual detriment that I could find!

File: packages/matrix/tests/skills.spec.ts [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+      (ev) =>
+        ev.type === 'm.room.message' &&
+        ev.content?.msgtype === 'org.boxel.message',
+    );
+    let { attachedCardsEventIds, attachedSkillEventIds } = JSON.parse(
+      messages[0]?.content?.data,
+    );
+    expect(attachedCardsEventIds).toHaveLength(1);
+    expect(attachedSkillEventIds).toHaveLength(2);
+  });
```
Comment (backspace) [inline packages/matrix/tests/skills.spec.ts:461]: nice tests 🤩

**Review by backspace — APPROVED** [review]
this worked pretty well locally other than that problem that’s surely more of an LLM problem than this code, I enjoy the pill animation 🎉

File: packages/base/room.gts (hunk 1 above)
Comment (tintinthong) [reply packages/base/room.gts:470]: I think `getCardEventIds` code in the matrix service is one of the only places with clashes. Particularly, we just deal with cardHashes in the service itself

**Review by tintinthong — APPROVED** [review]
Looks super cool

File: packages/base/skill-card.gts (hunk 2 above)
Comment (IanCal) [reply packages/base/skill-card.gts:25]: This should use the same mechanism for attaching cards, that is that the skill gets serialised and attached in the messages. If the serialisation is the same as previously, it points to the same event, otherwise it updates a new version.

I wonder if this is down to where gets updates? We can file a bug for this, and see if it happens similarly with attached cards.

Comment (burieberry) [reply packages/base/skill-card.gts:25]: when the skill is turned off, we remove it from the attached skills that's sent with the message. that's the slight difference from attached cards, which are cumulative. however, even though it's removed, AI still can reference it if it wants to, since it's part of the past room events, which we do not change, so turning a skill off is not 100% guaranteed. this was the main reason we only have a 'turn off' switch instead of a remove button for attached skills, since you can't really remove the knowledge fully once it's been attached, unless you start a new session.
